### DISCUSSION TITLE
how to transfer an allocated array on GPU to solver-apply

### DISCUSSION THREAD

**Discussion by ztdepztdep** [ginkgo-project/ginkgo#1132] [Q&A] [answered]
The d_b and d_x is two double arraies which have been allocated on GPU. I want to know how to transfer it to the solver-apply.
I tried to use the following code: 
     solver->apply(gko::lend(d_b), gko::lend(d_x));
but it doesn't wok.

**Comment by pratikvn** [comment]
Can you please let us know the following info: 
* Are `d_b` and `d_x` of type `gko::matrix::Dense<double>` ? 
* Can you elaborate on what errors you are getting ? Maybe add the error here below ?

**Comment by yhmtsai** [comment, marked as answer]
I copy my reply from the same discussion https://github.com/ginkgo-project/ginkgo/discussions/1131

Are the type of `d_b` and `d_x` `double*`?
If they are, the solver needs to use the `gko::matrix::Dense<double>*` type.
You can create the Dense matrix by reading a mtx file.
ex. `gko::read<gko::matrix::Dense<>>(std::ifstream("filepath"), exec);`
or create the Dense first then fill in the data
```
auto x = gko::matrix::Dense<>::create(exec, gko::dim<2>(n, 1));
// copy the data to x->get_values()
```
or create the Dense on the preallocated array
`auto x = gko::matrix::Dense<>::create(exec, gko::dim<2>(n, 1), gko::array<double>::view(exec, n, d_x));`
note. array view to indicate the size and location for the plain pointer.

**Reply by ztdepztdep** [reply]
how to copy the value to x->get_values(), need i to use clone?
the second command auto x = gko::matrix::Dense<>::create(exec, gko::dim<2>(n, 1), gko::array<double>::view(exec, n, d_x));
give many errors
/home/ztdep/lib/ginkgo/include/ginkgo/core/base/polymorphic_object.hpp(737): error: no instance of constructor "gko::matrix::Dense<ValueType>::Dense [with ValueType=double]" matches the argument list

**Reply by yhmtsai** [reply]
Sorry, I miss a parameter stride in the end. 
you can use `dest_exec->copy_from(src_exec, num_elems, src_ptr, dest_ptr)` to copy the value
